### PR TITLE
Added factories

### PR DESCRIPTION
**Pull request by ofrimasad**
- Added factories
- connected DatasetsFactory and MetricsFactory to SgModel
- added an example for factory usage to cifar10_resnet.yaml and cifar10_resnet_train_params.yaml

TODO - use factories in all recipes (in next PR)

### CODE REVIEW

File: src/super_gradients/common/factories/datasets_factory.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            "library_dataset": LibraryDatasetInterface,
+            "classification_dataset": ClassificationDatasetInterface,
+            "cifar_10": Cifar10DatasetInterface,
+            "cifar_100": Cifar100DatasetInterface,
+            "imagenet": ImageNetDatasetInterface,
+            "tiny_imagenet": TinyImageNetDatasetInterface,
+            "coco2017_detection": CoCoDetectionDatasetInterface,
+            "coco2014_detection": CoCo2014DetectionDatasetInterface,
+            "coco2017_segmentation": CoCoSegmentationDatasetInterface,
+            "pascal_voc": PascalVOC2012SegmentationDataSetInterface,
```
Comment (shaydeci) [inline src/super_gradients/common/factories/datasets_factory.py:24]: We should state the exact dataset (i.e year, voc/aug, segmentation/detection) from pascal. Note that we also have PascalVOCUnifiedDetectionDataSetInterface

File: src/super_gradients/common/factories/metrics_factory.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+from super_gradients.training.metrics import Accuracy, Top5, DetectionMetrics
+
+
+class MetricsFactory(BaseFactory):
+
+    def __init__(self):
+        type_dict = {
+            'Accuracy': Accuracy,
+            'Top5': Top5,
+            'DetectionMetrics': DetectionMetrics,
```
Comment (shaydeci) [inline src/super_gradients/common/factories/metrics_factory.py:11]: We have more then these...IoU for example...

File: src/super_gradients/common/factories/base_factory.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    def get(self, conf: Union[str, dict]):
+        """
+        Get an instantiated object.
+            :param conf: a configuration. either a type name (str) or a dictionary {type_name(str): {parameters...}}
+        """
+        raise NotImplementedError
+
+
+class BaseFactory(AbstractFactory):
+
```
Comment (shaydeci) [inline src/super_gradients/common/factories/base_factory.py:18]: This is a bit hard to follow, can you please add documentation that's a bit more detailed ?
Also im not sure I understand properly what the behaviour would be for instantiated objects inside a list ( such as train_metrics_list = [Accuracy(), Top5()]).

It is important we don't limit our users to only use our built in objects...

**Review by shaydeci — CHANGES_REQUESTED** [review]
Inline.

Besides that, we need to be very careful not to break our recipes, so please add a unit test that verifies that the same factory calls in them do not crash.

File: src/super_gradients/common/factories/base_factory.py (hunk 3 above)
Comment (ofrimasad) [reply src/super_gradients/common/factories/base_factory.py]: added documentation.
list factory allows the user to generate lists as well (explained in docs)

File: src/super_gradients/common/factories/datasets_factory.py (hunk 1 above)
Comment (ofrimasad) [reply src/super_gradients/common/factories/datasets_factory.py]: fixed.
PascalVOCDetectionDataSet does not have an interface currently

File: src/super_gradients/common/factories/metrics_factory.py (hunk 2 above)
Comment (ofrimasad) [reply src/super_gradients/common/factories/metrics_factory.py:11]: added IoU and PixelAccuracy

**Review by ofrimasad — COMMENTED** [review]
addressed

File: src/super_gradients/common/factories/datasets_factory.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/common/factories/datasets_factory.py]: Sure it does, https://github.com/Deci-AI/super-gradients/blob/2aa4045d9236b45623de4d892dd7c767c8398566/src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py#L697

**Review by shaydeci — CHANGES_REQUESTED** [review]
Thanks. Can we just add a unit test, so we know this won't break our recipes?
Something that takes a  yaml file and checks that it can be instantiated using the factory.

**Review by ofrimasad — COMMENTED** [review]
added a test

File: src/super_gradients/common/factories/datasets_factory.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/common/factories/datasets_factory.py]: Also cityscapes
